addons/sale_timesheet/models/account_move.py

```diff
@@ -41,47 +42,69 @@ def _compute_timesheet_count(self):
-            'res_model': 'account.analytic.line',
-            'view_id': False,
-            'view_mode': 'list,form',
-            'help': _("""
-                <p class="o_view_nocontent_smiling_face">
-                    Record timesheets
-                </p><p>
-                    You can register and track your workings hours by project every
-                    day. Every time spent on a project will become a cost and can be re-invoiced to
-                    customers if required.
-                </p>
-            """),
-            'limit': 80,
-            'context': {
-                'default_project_id': self.id,
-                'search_default_project_id': [self.id]
-            }
-        }
-
-    def _link_timesheets_to_invoice(self, start_date=None, end_date=None):
+        action = self.env['ir.actions.actions']._for_xml_id('sale_timesheet.action_timesheet_from_invoice')
+        context = ast.literal_eval(action.get('context', '{}'))
+        context['default_is_so_line_edited'] = True
+
+        default_so_line = next((sale_line for sale_line in self.line_ids.sale_line_ids if sale_line.is_service and sale_line.product_id.service_policy != 'delivered_timesheet'), False)
+        if default_so_line:
+            context['default_so_line'] = default_so_line.id
+            context['default_timesheet_invoice_id'] = self.id
+            context['default_task_id'] = default_so_line.task_id.id
+            if default_so_line.project_id:
+                context['default_project_id'] = default_so_line.project_id.id
+            else:
+                context['default_project_id'] = default_so_line.task_id.project_id.id
+        action['context'] = context
+
+        return action
+
+    def _link_timesheets_to_invoice(self, start_date=None, end_date=None, service_policies=('delivered_timesheet')):
         """ Search timesheets from given period and link this timesheets to the invoice
 
             When we create an invoice from a sale order, we need to
             link the timesheets in this sale order to the invoice.
             Then, we can know which timesheets are invoiced in the sale order.
             :param start_date: the start date of the period
             :param end_date: the end date of the period
+            :param service_policies: the service policies of the sol products to consider
         """
-        for line in self.filtered(lambda i: i.move_type == 'out_invoice' and i.state == 'draft').invoice_line_ids:
-            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.invoice_policy == 'delivery' and sol.product_id.service_type == 'timesheet')
+        for line in self.filtered(lambda i: i.move_type in ('out_invoice', 'out_refund') and i.state == 'draft').invoice_line_ids:
+            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.service_policy and sol.product_id.service_policy in service_policies)
```

Comment:
```suggestion
            sale_line_delivery = line.sale_line_ids.filtered(lambda sol: sol.product_id.service_policy in service_policies)
```
why is the first condition necessary?

Reply:
I don't quite remember, but I guess we can remove it indeed :sweat_smile:

Reply:
Okay now I remember, it's because `sol.product_id.service_policy` can be Falsy, and as `service_policies` is a list of strings, it will produce `TypeError: 'in <string>' requires string as left operand, not bool`